Mathlib/Analysis/InnerProductSpace/JointEigenspace.lean

```diff
@@ -0,0 +1,108 @@
+/-
+Copyright (c) 2024 Jon Bannon. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jon Bannon, Jack Cheverton, Samyak Dhar Tuladhar
+-/
+
+import Mathlib.Analysis.InnerProductSpace.Spectrum
+import Mathlib.Analysis.InnerProductSpace.Projection
+import Mathlib.Order.CompleteLattice
+import Mathlib.LinearAlgebra.Eigenspace.Basic
+
+/-! # Joint eigenspaces of a commuting pair of symmetric operators
+
+This file collects various decomposition results for joint eigenspaces of a commuting pair
+of symmetric operators on a finite-dimensional inner product space.
+
+# Main Result
+
+* `LinearMap.IsSymmetric.directSum_isInternal_of_commute` establishes that
+   if `{A B : E →ₗ[𝕜] E}`, then `IsSymmetric A`, `IsSymmetric B` and `A ∘ₗ B = B ∘ₗ A` imply that
+   `E` decomposes as an internal direct sum of the pairwise orthogonal spaces
+   `eigenspace B μ ⊓ eigenspace A ν`
+
+## TODO
+
+Develop a `Diagonalization` structure for linear maps and / or matrices which consists of a basis,
+and a proof obligation that the basis vectors are eigenvectors.
+
+## Tags
+
+self-adjoint operator, simultaneous eigenspaces, joint eigenspaces
+
+-/
+
+variable {𝕜 E : Type*} [RCLike 𝕜]
+variable [NormedAddCommGroup E] [InnerProductSpace 𝕜 E] [FiniteDimensional 𝕜 E]
+
+open Module.End
+
+namespace LinearMap
+
+namespace IsSymmetric
+
+section Pair
+
+variable {α : 𝕜} {A B : E →ₗ[𝕜] E} (hA : A.IsSymmetric) (hB : B.IsSymmetric)
+    (hAB : A ∘ₗ B = B ∘ₗ A)
+
+/--If a pair of operators commute, then the eigenspaces of one are invariant under the other.-/
+theorem eigenspace_invariant_of_commute (α : 𝕜) :
+    ∀ v ∈ (eigenspace A α), (B v ∈ eigenspace A α) := by
```

Comment:
```suggestion
theorem eigenspace_invariant_of_commute (α : 𝕜) :
    ∀ v ∈ eigenspace A α, B v ∈ eigenspace A α := by
```

Reply:
Also, having explicit arguments in a variable command somewhere above (i.e. `hA`, `hB` and `hAB` in this case) is generally discouraged, because then we can't tell how to actually use the lemma just by looking at the lemma statement.